engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -196,10 +196,16 @@ public StrategyPriority canHandle(DataObject srcData, DataObject destData) {
     }
 
     private boolean isVolumeOnManagedStorage(VolumeInfo volumeInfo) {
```

Comment:
What about a unit test here?

Reply:
@DaanHoogland is planning on writing a unit test for the higher-level canHandle method (to add to this PR).

Reply:
That is why we should unit test methods instead. We need to reduce the test code. Otherwise it becomes almost impossible to maintain them.

Reply:
sorry, I'm missing your point

Reply:
I was under the impression that Mike said because you are going to develop a test case (unit-ish) for  `canHandle`, we do not need the unit test case for `isVolumeOnManagedStorage`. And that is a problem. We should test methods as a unit. The test method for `canHandle` should not depend on ` isVolumeOnManagedStorage `, we should mock its (`isVolumeOnManagedStorage`) implementation instead.

Also, the method `canHandle` is a little bit big, which can be a little problematic to test. Maybe if you extract the IFs body there (the ones with nested IFs) you would be able to more easily write the unit test for the `canHandle`.

Reply:
I don't want to test private methods. We can of course factor it out but is actually part of the unit. It is made a separate method for readabilities sake and not used anywhere else is it?

Reply:
What can I do then?! ;)

In my opinion big methods should always be broken into smaller methods (up to ~10 lines). Then, these smaller methods need clear and concise documentation. Moreover, all methods should always be unit-tested (no excuse, no exceptions). That is a way to save great deal of time and bugs in code addition and refactoring. Not to say that unit tests cost way less to run than integration tests.

Reply:
There is a lot of debate in the programmer community with regards to whether or not private methods should be explicitly tested or tested implicitly through a public-facing method.

I don’t think the CloudStack community has ever established its own standard with regards to this topic.

Reply:
ok, i am torn on this;
yes methods should be small fit easily in one screen and be easy to read.
yes every bit (or, if we are forgiving every line) of code should be unit tested.
testing private methods however required trickery that I don't like.

I created a test to test for a particular behaviour of canHandle() that comes down to isVolumeOnManageStorage() (@mike-tutkowski can you merge it?) in this case because I felt it was better then to test a trivial private method. Do we expect that not even the most trivial methods are tested implicitly but instead stubbed? We could draw a strict line but I'm for case by case evaluation. Code coverage in tests is more important to me.

Reply:
don't thank me yet. @mike-tutkowski has to first merge it in (t)his PR for you to review it and than be eternally grateful, @rafaelweingartner  ;)

Reply:
@mike-tutkowski are you going to add some unit test(s)? this is ready for merging otherwise.

Reply:
Also, I think @DaanHoogland was planning on adding a unit test, as well.

Reply:
@mike-tutkowski Daan has already sent a PR to add unit test to your branch, can you merge that. See https://github.com/mike-tutkowski/cloudstack/pull/6

Reply:
@mike-tutkowski the PR I said I wanted to create some unit tests is not this one, it is the other one I worked.